File: python/dnd-character/dnd_character.py
from random import randint

class Character:
    def __init__(self):
        self.strength = self.ability()
        self.dexterity = self.ability()
        self.constitution = self.ability()
        self.intelligence = self.ability()
        self.wisdom = self.ability()
        self.charisma = self.ability()
        self.hitpoints = 10 + modifier(self.constitution)

    def ability(self):
        return sum(
            sorted(
                [randint(1, 6), randint(1, 6), randint(1, 6), randint(1, 6)],
                reverse=True
            )[:3]
        )

def modifier(value):
    return (value - 10) // 2

File: nim/dnd-character/dnd_character.nim
import std/random
import std/algorithm
import std/sequtils
import std/math
  
type
  Character* = object
    strength*: int
    dexterity*: int
    constitution*: int
    intelligence*: int
    wisdom*: int
    charisma*: int
    hitpoints*: int

proc ability*: int =
  var l = [rand(1..6), rand(1..6), rand(1..6), rand(1..6)]
  l.sort()
  foldl(l[1..3], a + b)
  
  

proc modifier*(n: int): int =
  int(floor((n - 10) / 2))

proc initCharacter*: Character =
  var con = ability()
  Character(
            strength: ability(),
            dexterity: ability(),
            constitution: con,
            intelligence: ability(),
            wisdom: ability(),
            charisma: ability(),
            hitpoints: modifier(con) + 10
  )
